### ISSUE TITLE
Something wrong in deserialization of straightforward custom class

### ISSUE THREAD

**Issue by jpivarski** [scikit-hep/uproot5#880]
### Discussed in https://github.com/scikit-hep/uproot5/discussions/879

<div type='discussions-op-text'>

<sup>Originally posted by **mattiasoldani** April 17, 2023</sup>
Howdy!

I'm using uproot 4.3.7 to open a ROOT file (of which a sample is attached: [TEST230417_LOCAL_SMALL.zip](https://github.com/scikit-hep/uproot5/files/11249405/TEST230417_LOCAL_SMALL.zip)) and access a tree ("Z") that is inside it. One of the branches of this tree, "Cluster[6]", should be an array of 6 custom zCluster objects for each event. Each zCluster object includes several members, mostly float and TVector3 entries. I checked with ROOT and all the branch dimensions are as expected, as seen in the TBrowser screenshots below: the tree contains 116 events, whereas all the Cluster[6] numerical subbranches contain 116*6 = 696 entries.

![Z_Event_iEv](https://user-images.githubusercontent.com/62517661/232466982-a2d77f83-1c39-4ffc-a9e6-be7737ed16e1.jpg)
![Z_Event_Cluster_nHits](https://user-images.githubusercontent.com/62517661/232466977-79dfddea-fc19-4205-aaf6-ebdcabb56498.jpg)

If I try to open the same file in uproot via
```
file = uproot.open("TEST230417_LOCAL_SMALL.root")
tree = file["Z"]
```
I get the correct tree, i.e. `tree.num_entries` outputs 116 and `tree.show()` outputs
```
name                 | typename                 | interpretation                
---------------------+--------------------------+-------------------------------
Event                | zEvent                   | AsGroup(<TBranchElement 'Ev...
Event/TObject        | (group of fUniqueID:u... | AsGroup(<TBranchElement 'TO...
Event/TObject/fUn... | uint32_t                 | AsDtype('>u4')
Event/TObject/fBits  | uint32_t                 | AsDtype('>u4')
Event/iEv            | int32_t                  | AsDtype('>i4')
Event/W              | double                   | AsDtype('>f8')
Event/pK             | TLorentzVector           | AsStridedObjects(Model_TLor...
Event/xK             | TVector3                 | AsStridedObjects(Model_TVec...
Event/iPair          | int32_t                  | AsDtype('>i4')
Event/ig[6]          | int32_t[6]               | AsDtype("('>i4', (6,))")
Event/igClus[6]      | int32_t[6]               | AsDtype("('>i4', (6,))")
Event/pg[6]          | TLorentzVector           | AsStridedObjects(Model_TLor...
Event/Hit[6]         | zHit                     | AsStridedObjects(Model_zHit...
Event/nCluster       | int32_t                  | AsDtype('>i4')
Event/Cluster[6]     | zCluster                 | AsObjects(Model_zCluster)
Event/Vertex         | zVertex                  | AsGroup(<TBranchElement 'Ve...
Event/Vertex/Vert... | int32_t                  | AsDtype('>i4')
Event/Vertex/Vert... | int32_t[2]               | AsDtype("('>i4', (2,))")
Event/Vertex/Vert... | int32_t                  | AsDtype('>i4')
Event/Vertex/Vert... | TVector3                 | AsStridedObjects(Model_TVec...
Event/Vertex/Vert... | TLorentzVector           | AsStridedObjects(Model_TLor...
Event/Vertex/Vert... | TVector3                 | AsStridedObjects(Model_TVec...
Event/Vertex/Vert... | TLorentzVector           | AsStridedObjects(Model_TLor...
Event/Vertex/Vert... | int32_t                  | AsDtype('>i4')
```

However, `tree.arrays(["Event/Cluster[6]"])` raises a `CannotBeAwkward: classes that can contain members of the same type cannot be Awkward Arrays because the depth of instances is unbounded` error and `tree.arrays(["Event/Cluster[6]"], library="np")` outputs a dictionary of 116 entries whose values are single zCluster objects, i.e. one single zCluster per event. Accessing any of them via e.g. `tree["Event/Cluster[6]"].arrays(library="np")["Cluster[6]"][0].all_members` seems to correctly output all the information regarding one single cluster, i.e.
```
{'Status': 1,
 'Bin': 3,
 'iSubdet': 0,
 'Detected': 0,
 'Primary': 0,
 'Used': 0,
 'Charge': 1,
 'Eff': 1.0,
 'xTrue': <TVector3 (version 3) at 0x7feb5830dc40>,
 'xRec': <TVector3 (version 3) at 0x7feb58314280>,
 'PosRes': 0.015788506169338613,
 'xTrue1': <TVector3 (version 3) at 0x7feb58314190>,
 'xTrue2': <TVector3 (version 3) at 0x7feb58314040>,
 'xPre1': <TVector3 (version 3) at 0x7feb58314490>,
 'xPre2': <TVector3 (version 3) at 0x7feb583145b0>,
 'PreRes': 0.06397004001907128,
 'ETrue': 0.07086701659225284,
 'ERec': 0.03220946465343243,
 'ERes': 0.09040274561889668,
 'Converted': 1,
 'dXdZTrue': 0.13310240197699721,
 'dYdZTrue': -0.34894589067080145,
 'nHits': 1,
 'HitList': array([0, 1, 1, 1, 1, 1], dtype=int32),
 'nConverted': 0}
```
but where are the other 5 clusters? Am I missing something trivial?

Thank you very much!

</div>

**Comment by ioanaif** [comment]
Hi!  @mattiasoldani 

I investigated this and found out that the issue comes from the fact that when we build the breadcrumbs for keeping a record of the recursion, in this particular case, at the level of the `zHit` it tries to add `TVector3` multiple times to the breadcrumbs (for `xTrue`, then for `xTrue1`, `xTrue2` etc) which raises the error you see.   

I am working towards fixing this issue such that the recursion breadcrumbs are built correctly. 

On the second part of the issue, regarding the number of clusters that get outputted, I opened the file with the jsroot browser and it shows only 116 entries, the same as what uproot shows:

<img width="1438" alt="image" src="https://user-images.githubusercontent.com/9751871/235666542-5732a56f-559b-40dc-ab78-e5e6ac51d0b1.png">

**Comment by mattiasoldani** [comment]
Dear @ioanaif,

thank you very much for the update!

I'd only like to point out that it is correct that "nCluster" has 116 entries, i.e., as many as the tree entries. That is just a counter of, say, the clusters of interest in each event. On the other hand, "Cluster[6]" should contain six clusters per event, corresponding to a total of 116*6 = 696 entries in the whole array. All this is confirmed by inspecting the file with plain ROOT and the TBrowser, and I think it should be the same using JSROOT. On the other hand, opening the file with uproot leads to an array of 116 clusters, only 1 per event.

**Comment by ioanaif** [comment]
Dear @mattiasoldani 

Indeed, "Cluster[6]" should contain six clusters per event, I was mistaken in my previous comment. 

Thus, you uncovered 2 issues here: the first one (recursion issue) was fixed in #886, and for the second one (six cluster vs one cluster contained) I am working towards a solution to have uproot output the correct number of clusters.
